src/methods/join.js

```diff
@@ -286,67 +286,107 @@ export class JoinImplementation {
     // effects collected after pnc was constructed
     e: Effects
   ): ForkedAbruptCompletion {
+    let recurse = (x: PossiblyNormalCompletion, xe: Effects): [ForkedAbruptCompletion, Effects] => {
+      let nx = this.replacePossiblyNormalCompletionWithForkedAbruptCompletion(realm, x, ac, e);
+      let nxe = new Effects(nx, xe.generator, xe.modifiedBindings, xe.modifiedProperties, xe.createdObjects);
+      return [nx, nxe];
+    };
+
+    let cloneEffects = () => {
+      // TODO (hermanv) if we use e as is, it ends up being applied twice when the join of the normal
+      // path is applied to the current state. Follow up with a PR that allows Effects to get deeply cloned
+      // and then clone e at this point.
+      e = construct_empty_effects(
+        realm,
+        e.result instanceof Completion ? new ReturnCompletion(e.result.value) : e.result
+      );
+      // Note that ac may depend on the effects in the original e, so use e.result until the cloning is done.
+      ac = (e.result: any);
+    };
+
     // set up e with ac as the completion. It's OK to do this repeatedly since ac is not changed by recursive calls.
     e.result = ac;
+
+    // # match (pncc, pnca)
     let pncc = pnc.consequent;
+    let pnca = pnc.alternate;
+
+    // * case (AbruptCompletion, SimpleNormalCompletion)
+    // * case (AbruptCompletion, PossiblyNormalCompletion)
     if (pncc instanceof AbruptCompletion) {
-      e = realm.composeEffects(pnc.alternateEffects, e);
-      if (pnc.alternate instanceof SimpleNormalCompletion) {
+      if (pnca instanceof SimpleNormalCompletion) {
+        e = realm.composeEffects(pnc.alternateEffects, e);
         return new ForkedAbruptCompletion(realm, pnc.joinCondition, pncc, pnc.consequentEffects, ac, e);
       }
-      invariant(pnc.alternate instanceof PossiblyNormalCompletion);
-      let na = this.replacePossiblyNormalCompletionWithForkedAbruptCompletion(realm, pnc.alternate, ac, e);
-      let ae = pnc.alternateEffects;
-      let nae = new Effects(na, ae.generator, ae.modifiedBindings, ae.modifiedProperties, ae.createdObjects);
+      invariant(pnca instanceof PossiblyNormalCompletion);
+      let [na, nae] = recurse(pnca, pnc.alternateEffects);
```

Comment:
It took me a while to figure out that recurse captures and e and sneaks it into nae. I would make it more explicit if it were up to me.

Reply:
👍